core/array.cpp

```diff
@@ -108,12 +112,34 @@ uint32_t Array::hash() const {
 	}
 	return h;
 }
-void Array::operator=(const Array &p_array) {
 
-	_ref(p_array);
+void Array::_assign(const Array &p_array) {
+
+	if (_p->typed.type != Variant::OBJECT && _p->typed.type == p_array._p->typed.type) {
+		//same type or untyped, just reference, shuold be fine
+		_ref(p_array);
+	} else if (_p->typed.type == Variant::NIL) { //from typed to untyped, must copy, but this is cheap anyway
+		_p->array = p_array._p->array;
```

Comment:
Same as below, this assignment should not be allowed. If typed to untyped is desired, then it should be done explicit with clone/duplicate.

Reply:
This is also fine imo. If not done a lot of things in the engine will just not work. Again, remember that this is intended for GDScript and other interpreted languages, where if users don't want to have anything to do with typing, they just won't. Again, you don't have to allow it from the C# side.

Reply:
All that really matters in this context, and the reason these checks exist,  is that a typed array can't get incorrect types inside. For anything else, users don't really care.